core/src/services/fs/writer.rs

```diff
@@ -101,3 +109,77 @@ impl oio::BlockingWrite for FsWriter<std::fs::File> {
         Ok(())
     }
 }
+
+impl oio::PositionWrite for FsWriter<tokio::fs::File> {
+    async fn write_all_at(&self, offset: u64, buf: Buffer) -> Result<()> {
+        let f = self.f.as_ref().expect("FsWriter must be initialized");
+
+        let f = f
+            .try_clone()
+            .await
+            .map_err(new_std_io_error)?
+            .into_std()
+            .await;
+
+        tokio::task::spawn_blocking(move || {
+            let mut buf = buf;
+            let mut offset = offset;
+            while !buf.is_empty() {
+                match write_at(&f, buf.chunk(), offset) {
+                    Ok(n) => {
+                        buf.advance(n);
+                        offset += n as u64
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+            Ok(())
+        })
+        .await
+        .map_err(new_task_join_error)?
+    }
+
+    async fn close(&self) -> Result<()> {
+        let f = self.f.as_ref().expect("FsWriter must be initialized");
+
+        let mut f = f
+            .try_clone()
+            .await
+            .map_err(new_std_io_error)?
+            .into_std()
+            .await;
+
+        f.flush().map_err(new_std_io_error)?;
+        f.sync_all().map_err(new_std_io_error)?;
+
+        if let Some(tmp_path) = &self.tmp_path {
+            tokio::fs::rename(tmp_path, &self.target_path)
+                .await
+                .map_err(new_std_io_error)?;
+        }
+        Ok(())
+    }
+
+    async fn abort(&self) -> Result<()> {
+        if let Some(tmp_path) = &self.tmp_path {
+            tokio::fs::remove_file(tmp_path)
+                .await
+                .map_err(new_std_io_error)
+        } else {
+            Err(Error::new(
+                ErrorKind::Unsupported,
+                "Fs doesn't support abort if atomic_write_dir is not set",
+            ))
+        }
+    }
+}
+
+#[cfg(target_os = "windows")]
+fn write_at(f: &File, buf: &[u8], offset: u64) -> Result<usize> {
+    f.seek_write(buf, offset).map_err(new_std_io_error)
+}
+
+#[cfg(not(target_os = "windows"))]
```

Comment:
Please use `cfg(unix)` and `cfg(windows)`

Reply:
fixed.